src/main/UserActivityMonitor.js

```diff
@@ -0,0 +1,157 @@
+// Copyright (c) 2015-2016 Yuya Ochiai
+// Copyright (c) 2016-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import EventEmitter from 'events';
+
+import electron from 'electron';
+
+const {app} = electron;
+
+/**
+ * Monitors system idle time, listens for system events and fires status updates as needed
+ */
+export default class UserActivityMonitor extends EventEmitter {
+  constructor() {
+    super();
+
+    this.isActive = true;
+    this.forceInactive = false;
+    this.idleTime = 0;
+    this.lastStatusUpdate = 0;
+    this.systemIdleTimeIntervalID = -1;
+
+    this.config = {
+      internalUpdateFrequencyMs: 1 * 1000,
+      statusUpdateThresholdMs: 60 * 1000,
+      activityTimeoutSec: 5 * 60,
+    };
+
+    // NOTE: binding needed to prevent error; fat arrow class methods don't work in current setup
+    // Error: "Error: async hook stack has become corrupted (actual: #, expected: #)"
+    this.handleSystemGoingAway = this.handleSystemGoingAway.bind(this);
+    this.handleSystemComingBack = this.handleSystemComingBack.bind(this);
+  }
+
+  get userIsActive() {
+    return this.forceInactive ? false : this.isActive;
+  }
+
+  get userIdleTime() {
+    return this.idleTime;
+  }
+
+  /**
+   * Begin monitoring system events and idle time at defined frequency
+   *
+   * @param {Object} config - overide internal configuration defaults
+   * @param {nunber} config.internalUpdateFrequencyMs
+   * @param {nunber} config.statusUpdateThresholdMs
+   * @param {nunber} config.activityTimeoutSec
+   * @emits {error} emitted when method is called before the app is ready
+   * @emits {error} emitted when this method has previously been called but not subsequently stopped
+   */
+  startMonitoring(config = {}) {
+    if (!app.isReady()) {
+      this.emit('error', new Error('UserActivityMonitor.startMonitoring can only be called after app is ready'));
+      return;
+    }
+
+    if (this.systemIdleTimeIntervalID >= 0) {
+      this.emit('error', new Error('User activity monitoring is already in progress'));
+      return;
+    }
+
+    this.config = Object.assign({}, this.config, config);
+
+    // NOTE: electron.powerMonitor cannot be referenced until the app is ready
```

Comment:
Hi. I am looking forward to releasing this feature.
As stated in the comments, isn't it necessary?
In my environment, I get an error without it.
```
// NOTE: electron.powerMonitor can not be referenced until the app is ready
electron.app.on ('ready', () => {
  electron.powerMonitor.on('suspend', this.handleSuspend);
  ...
})
```

Reply:
Got the app running locally now in the background while I work the new PR for this and I'm staying active the whole time! 🙌

Yes, you are absolutely correct! 🙂 I put the comment in there to remind that the _startMonitoring_ method needs to be called once the app is ready because of the use of _electron.powerMonitor_ or you get the error you mentioned. In _main.js_, the _startMonitoring_ method is called after the app is ready. If the method happens to be called before the app is ready, there is a conditional at the top of the method to emit an error and return without starting the monitoring.